Importando as bibliotecas necessárias

In [22]:

import pandas as pd
import numpy as np
import plotly.express as px
from sklearn.preprocessing import StandardScaler, MinMaxScaler, LabelEncoder
import io
import os


In [23]:
# Caminho para o dataset
dataset_path = '../data/ACC_INTAKES_OUTCOMES.parquet'

# Carregando os dados
data = pd.read_parquet(dataset_path)

# Exibindo as primeiras linhas do dataset para verificar se está tudo certo
data.head()



age_upon_outcome animal_id_outcome     date_of_birth outcome_subtype  \
0         10 years           A047759  02/04/2004 00:00       Partner     
1         14 years           A197810  21/01/2000 00:00       Partner     
2         16 years           A200922  03/10/1997 00:00       Foster      
3         14 years           A208755  27/04/2000 00:00       Suffering   
4         17 years           A210457  01/06/1999 00:00       Foster      

      outcome_type sex_upon_outcome  age_upon_outcome_(days)  \
0  Transfer           Neutered Male                     3650   
1  Transfer           Spayed Female                     5110   
2  Adoption           Neutered Male                     5840   
3  Euthanasia         Neutered Male                     5110   
4  Adoption           Neutered Male                     6205   

   age_upon_outcome_(years) age_upon_outcome_age_group  outcome_datetime  ...  \
0                      10.0               (7.5, 10.0]   07/04/2014 15:12  ...   
1                      14.0               (12.5, 15.0]  22/12/2014 15:23  ...   
2                      16.0               (15.0, 17.5]  22/11/2013 09:44  ...   
3                      14.0               (12.5, 15.0]  17/02/2015 16:25  ...   
4                      17.0               (15.0, 17.5]  07/10/2016 12:34  ...   

   age_upon_intake_age_group   intake_datetime intake_month intake_year  \
0               (7.5, 10.0]   02/04/2014 15:55            4        2014   
1               (12.5, 15.0]  08/12/2014 12:30           12        2014   
2               (15.0, 17.5]  03/10/2013 15:47           10        2013   
3               (12.5, 15.0]  17/02/2015 13:17            2        2015   
4               (15.0, 17.5]  28/09/2016 12:05            9        2016   

   intake_monthyear  intake_weekday  intake_hour  intake_number  \
0           2014-04       Wednesday           15              1   
1           2014-12       Monday              12              1   
2           2013-10       Thursday            15              1   
3           2015-02       Tuesday             13              1   
4           2016-09       Wednesday           12              1   

               time_in_shelter time_in_shelter_days  
0  4 days 23:17:00.000000000                    5.0  
1  14 days 02:53:00.000000000                  14.1  
2  49 days 17:57:00.000000000                  49.7  
3  0 days 03:08:00.000000000                    0.1  
4  9 days 00:29:00.000000000                    9.0  

[5 rows x 41 columns]

In [24]:
# Verificar informações básicas do dataset
data.info()



<class 'pandas.core.frame.DataFrame'>
RangeIndex: 36346 entries, 0 to 36345
Data columns (total 41 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   age_upon_outcome            36346 non-null  object 
 1   animal_id_outcome           36346 non-null  object 
 2   date_of_birth               36346 non-null  object 
 3   outcome_subtype             36346 non-null  object 
 4   outcome_type                36346 non-null  object 
 5   sex_upon_outcome            36346 non-null  object 
 6   age_upon_outcome_(days)     36346 non-null  int64  
 7   age_upon_outcome_(years)    36346 non-null  float64
 8   age_upon_outcome_age_group  36346 non-null  object 
 9   outcome_datetime            36346 non-null  object 
 10  outcome_month               36346 non-null  int64  
 11  outcome_year                36346 non-null  int64  
 12  outcome_monthyear           36346 non-null  object 
 13  outcome_weekday             363

In [25]:
# Exibir estatísticas descritivas do dataset
data.describe(include='all')


age_upon_outcome animal_id_outcome     date_of_birth outcome_subtype  \
count             36346             36346             36346           36346   
unique               46             35986              4838              19   
top            1 year             A695798  01/05/2016 00:00       Partner     
freq               6440                 4                74           19840   
mean                NaN               NaN               NaN             NaN   
std                 NaN               NaN               NaN             NaN   
min                 NaN               NaN               NaN             NaN   
25%                 NaN               NaN               NaN             NaN   
50%                 NaN               NaN               NaN             NaN   
75%                 NaN               NaN               NaN             NaN   
max                 NaN               NaN               NaN             NaN   

           outcome_type sex_upon_outcome  age_upon_outcome_(days)  \
count             36346            36346             36346.000000   
unique                6                5                      NaN   
top     Transfer           Neutered Male                      NaN   
freq              23799             8349                      NaN   
mean                NaN              NaN               636.381060   
std                 NaN              NaN               997.432904   
min                 NaN              NaN                 0.000000   
25%                 NaN              NaN                60.000000   
50%                 NaN              NaN               330.000000   
75%                 NaN              NaN               730.000000   
max                 NaN              NaN              9125.000000   

        age_upon_outcome_(years) age_upon_outcome_age_group  outcome_datetime  \
count               36346.000000                      36346             36346   
unique                       NaN                         10             26037   
top                          NaN               (-0.025, 2.5  18/04/2016 00:00   
freq                         NaN                      29355                39   
mean                    1.745694                        NaN               NaN   
std                     2.731384                        NaN               NaN   
min                     0.000000                        NaN               NaN   
25%                     0.200000                        NaN               NaN   
50%                     0.900000                        NaN               NaN   
75%                     2.000000                        NaN               NaN   
max                    25.000000                        NaN               NaN   

        ...  age_upon_intake_age_group   intake_datetime  intake_month  \
count   ...                      36346             36346  36346.000000   
unique  ...                         10             25148           NaN   
top     ...               (-0.025, 2.5  09/07/2014 12:58           NaN   
freq    ...                      29437                62           NaN   
mean    ...                        NaN               NaN      6.609696   
std     ...                        NaN               NaN      3.214465   
min     ...                        NaN               NaN      1.000000   
25%     ...                        NaN               NaN      4.000000   
50%     ...                        NaN               NaN      6.000000   
75%     ...                        NaN               NaN      9.000000   
max     ...                        NaN               NaN     12.000000   

         intake_year  intake_monthyear  intake_weekday   intake_hour  \
count   36346.000000             36346           36346  36346.000000   
unique           NaN                54               7           NaN   
top              NaN           2015-06       Monday              NaN   
freq             NaN              1246            5718           NaN   
mean   

In [26]:
data_cleaned = data.copy()

# Converter colunas de datas e horas para o tipo datetime com formato específico
date_columns = ['date_of_birth', 'outcome_datetime', 'intake_datetime']

for col in date_columns:
    data_cleaned[col] = pd.to_datetime(data_cleaned[col], format='%d/%m/%Y %H:%M', errors='coerce')

# Converter colunas 'monthyear' para o formato mm/yyyy
monthyear_columns = ['dob_monthyear', 'outcome_monthyear', 'intake_monthyear']
for col in monthyear_columns:
    data_cleaned[col] = pd.to_datetime(data_cleaned[col], format='%Y-%m', errors='coerce').dt.strftime('%m/%Y')

# Tratar colunas de mês como números
month_columns = ['outcome_month', 'dob_month', 'intake_month']
for col in month_columns:
    data_cleaned[col] = pd.to_numeric(data_cleaned[col], errors='coerce')

# Converter colunas numéricas para tipos numéricos
numeric_columns = [
    'age_upon_outcome_(days)', 'age_upon_outcome_(years)', 'outcome_year', 'outcome_hour',
    'outcome_number', 'dob_year', 'age_upon_intake_(days)', 'age_upon_intake_(years)',
    'intake_year', 'intake_hour', 'intake_number', 'time_in_shelter_days'
]
for col in numeric_columns:
    data_cleaned[col] = pd.to_numeric(data_cleaned[col], errors='coerce')

# Exibir a nova estrutura do DataFrame após as conversões
data_cleaned.info()



<class 'pandas.core.frame.DataFrame'>
RangeIndex: 36346 entries, 0 to 36345
Data columns (total 41 columns):
 #   Column                      Non-Null Count  Dtype         
---  ------                      --------------  -----         
 0   age_upon_outcome            36346 non-null  object        
 1   animal_id_outcome           36346 non-null  object        
 2   date_of_birth               36346 non-null  datetime64[ns]
 3   outcome_subtype             36346 non-null  object        
 4   outcome_type                36346 non-null  object        
 5   sex_upon_outcome            36346 non-null  object        
 6   age_upon_outcome_(days)     36346 non-null  int64         
 7   age_upon_outcome_(years)    36346 non-null  float64       
 8   age_upon_outcome_age_group  36346 non-null  object        
 9   outcome_datetime            36346 non-null  datetime64[ns]
 10  outcome_month               36346 non-null  int64         
 11  outcome_year                36346 non-null  int64     

In [27]:
# Verificar valores ausentes
missing_values = data.isnull().sum()
missing_values = missing_values[missing_values > 0].sort_values(ascending=False)
missing_values


Series([], dtype: int64)

In [28]:
# Verificar duplicatas
duplicate_count = data.duplicated().sum()
duplicate_count



15

In [29]:
# Remover duplicatas
data_cleaned = data.drop_duplicates()

# Exibir a contagem de registros após a remoção de duplicatas
data_cleaned.shape


(36331, 41)

In [30]:
# Função para identificar outliers
def identify_outliers(data):
    numeric_columns = data.select_dtypes(include=['float64', 'int64']).columns
    outlier_columns = []
    for col in numeric_columns:
        Q1 = data[col].quantile(0.25)
        Q3 = data[col].quantile(0.75)
        IQR = Q3 - Q1
        outlier_step = 1.5 * IQR
        if data[(data[col] < Q1 - outlier_step) | (data[col] > Q3 + outlier_step)].shape[0] > 0:
            outlier_columns.append(col)
    return outlier_columns

# Identificar as colunas com outliers
outlier_columns = identify_outliers(data_cleaned)

# Exibir as colunas que contêm outliers
if outlier_columns:
    print(f"Outliers identificados nas seguintes colunas: {', '.join(outlier_columns)}")
else:
    print("Não foram identificados outliers nas colunas numéricas.")


Outliers identificados nas seguintes colunas: age_upon_outcome_(days), age_upon_outcome_(years), outcome_hour, outcome_number, dob_year, age_upon_intake_(days), age_upon_intake_(years), intake_hour, intake_number, time_in_shelter_days


Como colunas relacionadas à idade e tempo no abrigo são necessárias para a analise de dados deste dataset, serão removidos apenas os outliers relacionados à horário de saída e entrada no abrigo pois são irrelevantes para nosso estudo.

In [31]:
# Função para remover outliers apenas nas colunas especificadas
def remove_outliers(data, columns):
    for col in columns:
        Q1 = data[col].quantile(0.25)
        Q3 = data[col].quantile(0.75)
        IQR = Q3 - Q1
        outlier_step = 1.5 * IQR
        data = data[~((data[col] < (Q1 - outlier_step)) | (data[col] > (Q3 + outlier_step)))]
    return data

# Colunas das quais removeremos os outliers
columns_to_remove_outliers = ['outcome_hour', 'intake_hour']

# Remover outliers dessas colunas
data_cleaned = remove_outliers(data, columns_to_remove_outliers)


# Exibir o número de registros restantes após a remoção de outliers
print(f"Número de registros restantes após remoção de outliers: {len(data_cleaned)}")


Número de registros restantes após remoção de outliers: 34358


Salvando dataset limpo

In [32]:
# Salvar como arquivo Parquet
data_cleaned.to_parquet('../data/data_cleaned.parquet', engine='pyarrow')

Escalonamento de dados

In [33]:
# Identificar todas as colunas numéricas no DataFrame
numeric_columns = data_cleaned.select_dtypes(include=['float64', 'int64']).columns

# Exibir as colunas numéricas para revisão
print("Colunas numéricas identificadas:")
print(numeric_columns)

Colunas numéricas identificadas:
Index(['age_upon_outcome_(days)', 'age_upon_outcome_(years)', 'outcome_month',
       'outcome_year', 'outcome_hour', 'outcome_number', 'dob_year',
       'dob_month', 'count', 'age_upon_intake_(days)',
       'age_upon_intake_(years)', 'intake_month', 'intake_year', 'intake_hour',
       'intake_number', 'time_in_shelter_days'],
      dtype='object')


Algumas colunas não serão escalonadas por serem irrelevantes para nosso estudo ou por já possuir outras colunas que cumprirão o proposito.

Neste caso, apenas as seguintes colunas serão escalonadas:

'dob_year' = Ano de nascimento do animal
'age_upon_intake_(days)' = Idade no momento de entrada em dias
'age_upon_intake_(years)' = Idade no momento de entrada em anos 
'intake_month', =  Mês de entrada
'intake_year', = Mês de saída 
'intake_number', = Número de vezes que já entrou no abrigo
'age_upon_outcome_(days)' = Idade no momento de saída em dias
'age_upon_outcome_(years)' = Idade no momento de saída em anos
'outcome_month' = Mês de saída
'outcome_year' = Ano de saída
'outcome_number' = Número de vezes que já saiu do abrigo
'time_in_shelter_days' = Quantidade de dias que passou no abrigo


Padronização dos dados (Standard Scaling):

Transforma os dados para que tenham média zero e desvio padrão igual a um. - É útil para algoritmos que assumem que os dados são distribuídos em uma forma normal (gaussiana).

In [34]:
from sklearn.preprocessing import StandardScaler

# Definir as colunas que precisam de escalonamento padronizado
columns_to_scale = [
'dob_year', 
'age_upon_intake_(days)',
'age_upon_intake_(years)',
'intake_month', 
'intake_year',  
'intake_number', 
'age_upon_outcome_(days)',
'age_upon_outcome_(years)',
'outcome_month',
'outcome_year',
'outcome_number', 
'time_in_shelter_days', 
]

# Inicializar o StandardScaler
scaler = StandardScaler()

# Aplicar o escalonamento padronizado e adicionar as colunas escalonadas ao DataFrame
data_scaled = data_cleaned.copy()  # Criar uma cópia para manter os dados originais
for col in columns_to_scale:
    data_scaled[f'{col}_scaled'] = scaler.fit_transform(data_cleaned[[col]])

# Verificar o resultado
data_scaled.head()


age_upon_outcome animal_id_outcome     date_of_birth outcome_subtype  \
0         10 years           A047759  02/04/2004 00:00       Partner     
1         14 years           A197810  21/01/2000 00:00       Partner     
2         16 years           A200922  03/10/1997 00:00       Foster      
3         14 years           A208755  27/04/2000 00:00       Suffering   
4         17 years           A210457  01/06/1999 00:00       Foster      

      outcome_type sex_upon_outcome  age_upon_outcome_(days)  \
0  Transfer           Neutered Male                     3650   
1  Transfer           Spayed Female                     5110   
2  Adoption           Neutered Male                     5840   
3  Euthanasia         Neutered Male                     5110   
4  Adoption           Neutered Male                     6205   

   age_upon_outcome_(years) age_upon_outcome_age_group  outcome_datetime  ...  \
0                      10.0               (7.5, 10.0]   07/04/2014 15:12  ...   
1                      14.0               (12.5, 15.0]  22/12/2014 15:23  ...   
2                      16.0               (15.0, 17.5]  22/11/2013 09:44  ...   
3                      14.0               (12.5, 15.0]  17/02/2015 16:25  ...   
4                      17.0               (15.0, 17.5]  07/10/2016 12:34  ...   

   age_upon_intake_(years)_scaled  intake_month_scaled intake_year_scaled  \
0                        3.035380            -0.813021          -1.009191   
1                        4.499287             1.677259          -1.009191   
2                        5.231241             1.054689          -1.788185   
3                        4.499287            -1.435591          -0.230197   
4                        5.597218             0.743404           0.548796   

  intake_number_scaled  age_upon_outcome_(days)_scaled  \
0            -0.145378                        3.004984   
1            -0.145378                        4.460630   
2            -0.145378                        5.188454   
3            -0.145378                        4.460630   
4            -0.145378                        5.552365   

   age_upon_outcome_(years)_scaled  outcome_month_scaled  outcome_year_scaled  \
0                         3.005601             -0.829338            -1.036047   
1                         4.461925              1.614592            -1.036047   
2                         5.190087              1.309101            -1.809951   
3                         4.461925             -1.440320            -0.262143   
4                         5.554168              1.003610             0.511761   

  outcome_number_scaled time_in_shelter_days_scaled  
0             -0.145378                   -0.303841  
1             -0.145378                   -0.113195  
2             -0.145378                    0.632628  
3             -0.145378                   -0.406496  
4             -0.145378                   -0.220041  

[5 rows x 53 columns]

Agora, realizaremos a Normalização dos dados.

Normalização (Min-Max Scaling):
Transforma os dados para que todos os valores estejam dentro do intervalo [0, 1].
É útil para algoritmos que não fazem suposições sobre a distribuição dos dados, como redes neurais e métodos de distância (por exemplo, k-NN).

In [35]:
from sklearn.preprocessing import MinMaxScaler

# Definir as colunas que precisam de normalização
columns_to_normalize = [
'dob_year', 
'age_upon_intake_(days)',
'age_upon_intake_(years)',
'intake_month', 
'intake_year',  
'intake_number', 
'age_upon_outcome_(days)',
'age_upon_outcome_(years)',
'outcome_month',
'outcome_year',
'outcome_number', 
'time_in_shelter_days', 
]

# Inicializar o MinMaxScaler
scaler = MinMaxScaler()

# Aplicar a normalização e adicionar as colunas normalizadas ao DataFrame
data_normalized = data_cleaned.copy()  # Criar uma cópia para manter os dados originais
for col in columns_to_normalize:
    data_normalized[f'{col}_normalized'] = scaler.fit_transform(data_cleaned[[col]])

# Verificar o resultado
data_normalized.head()


age_upon_outcome animal_id_outcome     date_of_birth outcome_subtype  \
0         10 years           A047759  02/04/2004 00:00       Partner     
1         14 years           A197810  21/01/2000 00:00       Partner     
2         16 years           A200922  03/10/1997 00:00       Foster      
3         14 years           A208755  27/04/2000 00:00       Suffering   
4         17 years           A210457  01/06/1999 00:00       Foster      

      outcome_type sex_upon_outcome  age_upon_outcome_(days)  \
0  Transfer           Neutered Male                     3650   
1  Transfer           Spayed Female                     5110   
2  Adoption           Neutered Male                     5840   
3  Euthanasia         Neutered Male                     5110   
4  Adoption           Neutered Male                     6205   

   age_upon_outcome_(years) age_upon_outcome_age_group  outcome_datetime  ...  \
0                      10.0               (7.5, 10.0]   07/04/2014 15:12  ...   
1                      14.0               (12.5, 15.0]  22/12/2014 15:23  ...   
2                      16.0               (15.0, 17.5]  22/11/2013 09:44  ...   
3                      14.0               (12.5, 15.0]  17/02/2015 16:25  ...   
4                      17.0               (15.0, 17.5]  07/10/2016 12:34  ...   

   age_upon_intake_(years)_normalized  intake_month_normalized  \
0                            0.454545                 0.272727   
1                            0.636364                 1.000000   
2                            0.727273                 0.818182   
3                            0.636364                 0.090909   
4                            0.772727                 0.727273   

  intake_year_normalized intake_number_normalized  \
0                    0.2                      0.0   
1                    0.2                      0.0   
2                    0.0                      0.0   
3                    0.4                      0.0   
4                    0.6                      0.0   

   age_upon_outcome_(days)_normalized  age_upon_outcome_(years)_normalized  \
0                            0.454545                             0.454545   
1                            0.636364                             0.636364   
2                            0.727273                             0.727273   
3                            0.636364                             0.636364   
4                            0.772727                             0.772727   

   outcome_month_normalized  outcome_year_normalized  \
0                  0.272727                      0.2   
1                  1.000000                      0.2   
2                  0.909091                      0.0   
3                  0.090909                      0.4   
4                  0.818182                      0.6   

  outcome_number_normalized time_in_shelter_days_normalized  
0                       0.0                        0.003984  
1                       0.0                        0.011235  
2                       0.0                        0.039602  
3                       0.0                        0.000080  
4                       0.0                        0.007171  

[5 rows x 53 columns]

Códificação dos dados + Feature Engineering

In [36]:
# Identificar colunas categóricas no dataset limpo
categorical_columns = data_cleaned.select_dtypes(include=['object', 'category']).columns

# Exibir a quantidade de valores únicos para cada coluna categórica
unique_values = data_cleaned[categorical_columns].nunique()

# Exibir o resultado
print("Quantidade de valores únicos para colunas categóricas:")
print(unique_values)


Quantidade de valores únicos para colunas categóricas:
age_upon_outcome                 45
animal_id_outcome             34030
date_of_birth                  4782
outcome_subtype                  19
outcome_type                      6
sex_upon_outcome                  5
age_upon_outcome_age_group        9
outcome_datetime              25303
outcome_monthyear                55
outcome_weekday                   7
dob_monthyear                    55
age_upon_intake                  45
animal_id_intake              34030
animal_type                       4
breed                          1170
color                           395
found_location                16377
intake_condition                  8
intake_type                       5
sex_upon_intake                   5
age_upon_intake_age_group         9
intake_datetime               23832
intake_monthyear                 54
intake_weekday                    7
time_in_shelter               16060
dtype: int64


Algumas colunas receberão One-Hot Encoding, outras Label Encoding e outras feature engineering.

One-Hot Encoding
Utilizamos One-Hot Encoding para colunas categóricas nominais onde não há uma ordem implícita entre os valores. Esta técnica cria uma nova coluna binária para cada valor único da categoria, permitindo que o modelo trate cada valor de maneira igual.

Label Encoding
Utilizamos Label Encoding para colunas categóricas ordinais onde há uma ordem implícita entre os valores. Esta técnica converte os valores categóricos em valores numéricos que refletem a ordem implícita entre eles.

Feature Engineering, é o processo de aplicar o conhecimento de domínio dos dados para criar recursos que ajudam os algoritmos de machine learning a aprender melhor.

Colunas que receberão One-Hot Encoding:
outcome_subtype                  
outcome_type                     
sex_upon_outcome                  
outcome_weekday                  
animal_type     
intake_condition                  
intake_type                       
sex_upon_intake                   
intake_weekday                    

In [37]:
# Lista das colunas categóricas a serem codificadas
categorical_columns = [
    'outcome_subtype', 
    'outcome_type', 
    'sex_upon_outcome', 
    'outcome_weekday', 
    'animal_type', 
    'intake_condition', 
    'intake_type', 
    'sex_upon_intake', 
    'intake_weekday'
]

# Aplicando One-Hot Encoding ao data_scaled
data_scaled = pd.get_dummies(data_scaled, columns=categorical_columns)

# Aplicando One-Hot Encoding ao data_normalized
data_normalized = pd.get_dummies(data_normalized, columns=categorical_columns)



In [38]:
# Verificando o resultado após o One-Hot Encoding no dataset Padronizado
data_scaled.head()


age_upon_outcome animal_id_outcome     date_of_birth  \
0         10 years           A047759  02/04/2004 00:00   
1         14 years           A197810  21/01/2000 00:00   
2         16 years           A200922  03/10/1997 00:00   
3         14 years           A208755  27/04/2000 00:00   
4         17 years           A210457  01/06/1999 00:00   

   age_upon_outcome_(days)  age_upon_outcome_(years)  \
0                     3650                      10.0   
1                     5110                      14.0   
2                     5840                      16.0   
3                     5110                      14.0   
4                     6205                      17.0   

  age_upon_outcome_age_group  outcome_datetime  outcome_month  outcome_year  \
0               (7.5, 10.0]   07/04/2014 15:12              4          2014   
1               (12.5, 15.0]  22/12/2014 15:23             12          2014   
2               (15.0, 17.5]  22/11/2013 09:44             11          2013   
3               (12.5, 15.0]  17/02/2015 16:25              2          2015   
4               (15.0, 17.5]  07/10/2016 12:34             10          2016   

  outcome_monthyear  ...  sex_upon_intake_Neutered Male  \
0           2014-04  ...                           True   
1           2014-12  ...                          False   
2           2013-11  ...                           True   
3           2015-02  ...                           True   
4           2016-10  ...                           True   

   sex_upon_intake_Spayed Female  sex_upon_intake_Unknown        \
0                          False                          False   
1                           True                          False   
2                          False                          False   
3                          False                          False   
4                          False                          False   

   intake_weekday_Friday    intake_weekday_Monday    intake_weekday_Saturday   \
0                     False                    False                    False   
1                     False                     True                    False   
2                     False                    False                    False   
3                     False                    False                    False   
4                     False                    False                    False   

  intake_weekday_Sunday    intake_weekday_Thursday  intake_weekday_Tuesday    \
0                    False                    False                    False   
1                    False                    False                    False   
2                    False                     True                    False   
3                    False                    False                     True   
4                    False                    False                    False   

  intake_weekday_Wednesday  
0                     True  
1                    False  
2                    False  
3                    False  
4                     True  

[5 rows x 110 columns]

In [39]:
# Verificando o resultado após o One-Hot Encoding no dataset Normalizado
data_normalized.head()


age_upon_outcome animal_id_outcome     date_of_birth  \
0         10 years           A047759  02/04/2004 00:00   
1         14 years           A197810  21/01/2000 00:00   
2         16 years           A200922  03/10/1997 00:00   
3         14 years           A208755  27/04/2000 00:00   
4         17 years           A210457  01/06/1999 00:00   

   age_upon_outcome_(days)  age_upon_outcome_(years)  \
0                     3650                      10.0   
1                     5110                      14.0   
2                     5840                      16.0   
3                     5110                      14.0   
4                     6205                      17.0   

  age_upon_outcome_age_group  outcome_datetime  outcome_month  outcome_year  \
0               (7.5, 10.0]   07/04/2014 15:12              4          2014   
1               (12.5, 15.0]  22/12/2014 15:23             12          2014   
2               (15.0, 17.5]  22/11/2013 09:44             11          2013   
3               (12.5, 15.0]  17/02/2015 16:25              2          2015   
4               (15.0, 17.5]  07/10/2016 12:34             10          2016   

  outcome_monthyear  ...  sex_upon_intake_Neutered Male  \
0           2014-04  ...                           True   
1           2014-12  ...                          False   
2           2013-11  ...                           True   
3           2015-02  ...                           True   
4           2016-10  ...                           True   

   sex_upon_intake_Spayed Female  sex_upon_intake_Unknown        \
0                          False                          False   
1                           True                          False   
2                          False                          False   
3                          False                          False   
4                          False                          False   

   intake_weekday_Friday    intake_weekday_Monday    intake_weekday_Saturday   \
0                     False                    False                    False   
1                     False                     True                    False   
2                     False                    False                    False   
3                     False                    False                    False   
4                     False                    False                    False   

  intake_weekday_Sunday    intake_weekday_Thursday  intake_weekday_Tuesday    \
0                    False                    False                    False   
1                    False                    False                    False   
2                    False                     True                    False   
3                    False                    False                     True   
4                    False                    False                    False   

  intake_weekday_Wednesday  
0                     True  
1                    False  
2                    False  
3                    False  
4                     True  

[5 rows x 110 columns]

Colunas que receberão Label Encoding:
age_upon_outcome_age_group        
age_upon_intake_age_group         

In [40]:
from sklearn.preprocessing import LabelEncoder

# Inicializar o LabelEncoder
label_encoder = LabelEncoder()

# Aplicar Label Encoding nas colunas especificadas
columns_to_label_encode = [
    'age_upon_outcome_age_group',
    'age_upon_intake_age_group'
]

# Aplicar Label Encoding para o dataset padronizado
data_scaled = data_scaled.copy()
for col in columns_to_label_encode:
    data_scaled[f'{col}_encoded'] = label_encoder.fit_transform(data_cleaned[col])

# Aplicar Label Encoding para o dataset normalizado
data_normalized = data_normalized.copy()
for col in columns_to_label_encode:
    data_normalized[f'{col}_encoded'] = label_encoder.fit_transform(data_cleaned[col])


In [41]:
# Verificando o resultado após o Label Encoding no dataset Padronizado
data_scaled.head()

age_upon_outcome animal_id_outcome     date_of_birth  \
0         10 years           A047759  02/04/2004 00:00   
1         14 years           A197810  21/01/2000 00:00   
2         16 years           A200922  03/10/1997 00:00   
3         14 years           A208755  27/04/2000 00:00   
4         17 years           A210457  01/06/1999 00:00   

   age_upon_outcome_(days)  age_upon_outcome_(years)  \
0                     3650                      10.0   
1                     5110                      14.0   
2                     5840                      16.0   
3                     5110                      14.0   
4                     6205                      17.0   

  age_upon_outcome_age_group  outcome_datetime  outcome_month  outcome_year  \
0               (7.5, 10.0]   07/04/2014 15:12              4          2014   
1               (12.5, 15.0]  22/12/2014 15:23             12          2014   
2               (15.0, 17.5]  22/11/2013 09:44             11          2013   
3               (12.5, 15.0]  17/02/2015 16:25              2          2015   
4               (15.0, 17.5]  07/10/2016 12:34             10          2016   

  outcome_monthyear  ...  sex_upon_intake_Unknown        \
0           2014-04  ...                          False   
1           2014-12  ...                          False   
2           2013-11  ...                          False   
3           2015-02  ...                          False   
4           2016-10  ...                          False   

   intake_weekday_Friday     intake_weekday_Monday     \
0                     False                     False   
1                     False                      True   
2                     False                     False   
3                     False                     False   
4                     False                     False   

   intake_weekday_Saturday  intake_weekday_Sunday    intake_weekday_Thursday   \
0                     False                    False                    False   
1                     False                    False                    False   
2                     False                    False                     True   
3                     False                    False                    False   
4                     False                    False                    False   

  intake_weekday_Tuesday   intake_weekday_Wednesday  \
0                    False                     True   
1                    False                    False   
2                    False                    False   
3                     True                    False   
4                    False                     True   

  age_upon_outcome_age_group_encoded age_upon_intake_age_group_encoded  
0                                  8                                 8  
1                                  2                                 2  
2                                  3                                 3  
3                                  2                                 2  
4                                  3                                 3  

[5 rows x 112 columns]

In [42]:
# Verificando o resultado após o Label Encoding no dataset Normalizado
data_normalized.head()

age_upon_outcome animal_id_outcome     date_of_birth  \
0         10 years           A047759  02/04/2004 00:00   
1         14 years           A197810  21/01/2000 00:00   
2         16 years           A200922  03/10/1997 00:00   
3         14 years           A208755  27/04/2000 00:00   
4         17 years           A210457  01/06/1999 00:00   

   age_upon_outcome_(days)  age_upon_outcome_(years)  \
0                     3650                      10.0   
1                     5110                      14.0   
2                     5840                      16.0   
3                     5110                      14.0   
4                     6205                      17.0   

  age_upon_outcome_age_group  outcome_datetime  outcome_month  outcome_year  \
0               (7.5, 10.0]   07/04/2014 15:12              4          2014   
1               (12.5, 15.0]  22/12/2014 15:23             12          2014   
2               (15.0, 17.5]  22/11/2013 09:44             11          2013   
3               (12.5, 15.0]  17/02/2015 16:25              2          2015   
4               (15.0, 17.5]  07/10/2016 12:34             10          2016   

  outcome_monthyear  ...  sex_upon_intake_Unknown        \
0           2014-04  ...                          False   
1           2014-12  ...                          False   
2           2013-11  ...                          False   
3           2015-02  ...                          False   
4           2016-10  ...                          False   

   intake_weekday_Friday     intake_weekday_Monday     \
0                     False                     False   
1                     False                      True   
2                     False                     False   
3                     False                     False   
4                     False                     False   

   intake_weekday_Saturday  intake_weekday_Sunday    intake_weekday_Thursday   \
0                     False                    False                    False   
1                     False                    False                    False   
2                     False                    False                     True   
3                     False                    False                    False   
4                     False                    False                    False   

  intake_weekday_Tuesday   intake_weekday_Wednesday  \
0                    False                     True   
1                    False                    False   
2                    False                    False   
3                     True                    False   
4                    False                     True   

  age_upon_outcome_age_group_encoded age_upon_intake_age_group_encoded  
0                                  8                                 8  
1                                  2                                 2  
2                                  3                                 3  
3                                  2                                 2  
4                                  3                                 3  

[5 rows x 112 columns]

As seguintes colunas receberão "Feature Engineering"
- breed
- color

Como a coluna "breed" possui 1170 valores únicos, criaremos uma nova coluna chamada "is_mix_breed" que será usada para identificar se o animal tem uma raça misturada ou pura. Esta adaptação será feita da seguinte forma: 

Se na coluna "breed" original o valor tiver "Mix" ou "/", a coluna "is_mix_breed" terá o valor 1. Se não, terá o valor 0.

In [43]:
# Categorização de raças e adição da coluna 'is_mix_breed' aos datasets escalonados e normalizados
def categorize_breeds(data):
    data['is_mix_breed'] = data['breed'].str.contains('Mix|/', case=False, na=False).astype(int)
    return data

# Adicionar a coluna 'is_mix_breed' aos datasets escalonados e normalizados
data_scaled['is_mix_breed'] = categorize_breeds(data_scaled)['is_mix_breed']
data_normalized['is_mix_breed'] = categorize_breeds(data_normalized)['is_mix_breed']


In [44]:
# Verificar o resultado no dataset Padronizado
data_scaled[['breed', 'is_mix_breed']].head()

breed  is_mix_breed
0  Dachshund                                         0
1  American Shorthair Mix                            1
2  Dachshund Mix                                     1
3  Border Collie Mix                                 1
4  Chihuahua Shorthair                               0

In [45]:
# Verificar o resultado no dataset Normalizado
data_normalized[['breed', 'is_mix_breed']].head()

breed  is_mix_breed
0  Dachshund                                         0
1  American Shorthair Mix                            1
2  Dachshund Mix                                     1
3  Border Collie Mix                                 1
4  Chihuahua Shorthair                               0

Feature Engineering na coluna "color": 

A coluna color do dataset apresenta valores que podem ser compostos por uma única cor ou por múltiplas cores separadas por uma barra (/). 

Para simplificar a análise e reduzir a quantidade de valores únicos nesta coluna, aplicamos os seguintes passos:

1) Processamento de Cores Múltiplas:

- As cores compostas por duas cores diferentes, como "Black/White", foram substituídas pelo valor "multi", indicando que o animal possui múltiplas cores.
- Cores que repetiam o mesmo valor, como "White/White", foram simplificadas para apenas "White".

2) Agrupamento por Categorias:

As cores serão categorizadas em grupos maiores (lista abaixo). Isso reduzirá significativamente o número de valores únicos na coluna, facilitando a análise.

Black
White
Gray/Blue
Brown/Chocolate
Yellow/Gold/Cream
Red/Orange
Patterned
Other_Colors

3) One-Hot Encoding:

A nova coluna categorizada, color_category, receberá One-Hot Encoding, transformando cada categoria em uma variável binária (0 ou 1). Essa transformação será aplicada tanto no dataset padronizado quanto no dataset normalizado.

In [51]:
# Função para processar a coluna 'color'
def process_color(color):
    if '/' in color:
        colors = color.split('/')
        if colors[0] == colors[1]:
            return colors[0].strip()
        else:
            return 'multi'
    return color.strip()

# Aplicar a função de processamento de cores nos datasets padronizados e normalizados
data_scaled['processed_color'] = data_scaled['color'].apply(process_color)
data_normalized['processed_color'] = data_normalized['color'].apply(process_color)

# Verificar a quantidade de valores únicos após o processamento
unique_colors_scaled = data_scaled['processed_color'].unique()
unique_colors_normalized = data_normalized['processed_color'].unique()


In [52]:
print(f"Quantidade de cores únicas (escalonado): {len(unique_colors_scaled)}")

Quantidade de cores únicas (escalonado): 56


In [53]:
print(f"Quantidade de cores únicas (normalizado): {len(unique_colors_normalized)}")

Quantidade de cores únicas (normalizado): 56


In [54]:
# Opcional: Exibir os valores únicos na nova coluna
print(data_normalized['processed_color'].unique())

['Tricolor' 'multi' 'Black' 'Blue Merle' 'Tan' 'White' 'Tortie' 'Cream'
 'Calico' 'Brown' 'Brown Tabby' 'Yellow' 'Blue Point' 'Silver Tabby' 'Red'
 'Brown Brindle' 'Chocolate' 'Lynx Point' 'Chocolate Point' 'Blue'
 'Blue Cream' 'Calico Point' 'Orange Tabby' 'Torbie' 'Gold' 'Black Smoke'
 'Buff' 'Red Merle' 'Black Brindle' 'Red Tick' 'Blue Tabby' 'Fawn' 'Sable'
 'Silver' 'Cream Tabby' 'Flame Point' 'Lilac Point' 'Blue Tick' 'Gray'
 'Tortie Point' 'Black Tabby' 'Seal Point' 'Apricot' 'Green' 'Gray Tabby'
 'Brown Merle' 'Brown Tiger' 'Orange' 'Pink' 'Yellow Brindle'
 'Silver Lynx Point' 'Agouti' 'Blue Tiger' 'Orange Tiger' 'Liver Tick'
 'Blue Smoke']


Apesar de termos diminuido a quantidade de valores únicos ainda há muitos valores. Então realizaremos a categorização dessas cores.

Categorias:
Black: 'Black', 'Black Smoke', 'Black Brindle', 'Black Tabby'
White: 'White', 'Cream', 'Cream Tabby', 'Silver', 'Silver Tabby', 'Silver Lynx Point'
Gray/Blue: 'Gray', 'Blue', 'Blue Tabby', 'Blue Smoke', 'Gray Tabby', 'Blue Cream', 'Blue Merle', 'Blue Point', 'Blue Tick'
Brown/Chocolate: 'Brown', 'Brown Tabby', 'Brown Brindle', 'Brown Merle', 'Brown Tiger', 'Chocolate', 'Chocolate Point', 'Liver Tick'
Yellow/Gold/Cream: 'Yellow', 'Yellow Brindle', 'Gold', 'Buff', 'Apricot', 'Tan', 'Fawn'
Red/Orange: 'Red', 'Red Merle', 'Red Tick', 'Flame Point', 'Orange', 'Orange Tabby', 'Orange Tiger'
Patterned: 'Tricolor', 'Torbie', 'Tortie', 'Tortie Point', 'Calico', 'Calico Point'
Other_Colors: 'multi', 'Green', 'Pink', 'Agouti', 'Lilac Point', 'Sable', 'Lynx Point', 'Seal Point'

In [56]:
# Função para categorizar cores
def categorize_color(color):
    color = color.strip()
    if color in ['Black', 'Black Smoke', 'Black Brindle', 'Black Tabby']:
        return 'Black'
    elif color in ['White', 'Cream', 'Cream Tabby', 'Silver', 'Silver Tabby', 'Silver Lynx Point']:
        return 'White'
    elif color in ['Gray', 'Blue', 'Blue Tabby', 'Blue Smoke', 'Gray Tabby', 'Blue Cream', 'Blue Merle', 'Blue Point', 'Blue Tick']:
        return 'Gray/Blue'
    elif color in ['Brown', 'Brown Tabby', 'Brown Brindle', 'Brown Merle', 'Brown Tiger', 'Chocolate', 'Chocolate Point', 'Liver Tick']:
        return 'Brown/Chocolate'
    elif color in ['Yellow', 'Yellow Brindle', 'Gold', 'Buff', 'Apricot', 'Tan', 'Fawn']:
        return 'Yellow/Gold/Cream'
    elif color in ['Red', 'Red Merle', 'Red Tick', 'Flame Point', 'Orange', 'Orange Tabby', 'Orange Tiger']:
        return 'Red/Orange'
    elif color in ['Tricolor', 'Torbie', 'Tortie', 'Tortie Point', 'Calico', 'Calico Point']:
        return 'Patterned'
    else:
        return 'Other_Colors'

# Aplicar a categorização de cores nos datasets padronizados e normalizados
data_scaled['color_category'] = data_scaled['processed_color'].apply(categorize_color)
data_normalized['color_category'] = data_normalized['processed_color'].apply(categorize_color)

# Verificar a quantidade de valores únicos após a categorização
unique_color_categories_scaled = data_scaled['color_category'].unique()
unique_color_categories_normalized = data_normalized['color_category'].unique()

print(f"Quantidade de categorias de cores únicas (scaled): {len(unique_color_categories_scaled)}")
print(f"Quantidade de categorias de cores únicas (normalized): {len(unique_color_categories_normalized)}")


Quantidade de categorias de cores únicas (scaled): 8
Quantidade de categorias de cores únicas (normalized): 8


Agora, iremos aplicar One-Hot Encoding na coluna de cores que foi categorizada.

In [57]:
# Aplicar One-Hot Encoding na coluna 'color_category'
data_scaled = pd.get_dummies(data_scaled, columns=['color_category'], prefix='color')
data_normalized = pd.get_dummies(data_normalized, columns=['color_category'], prefix='color')


In [58]:
#Verificar o resultado no dataset Padronizado
print("Colunas após One-Hot Encoding (scaled):")
print(data_scaled.columns)

Colunas após One-Hot Encoding (scaled):
Index(['age_upon_outcome', 'animal_id_outcome', 'date_of_birth',
       'age_upon_outcome_(days)', 'age_upon_outcome_(years)',
       'age_upon_outcome_age_group', 'outcome_datetime', 'outcome_month',
       'outcome_year', 'outcome_monthyear',
       ...
       'is_mix_breed', 'processed_color', 'color_Black',
       'color_Brown/Chocolate', 'color_Gray/Blue', 'color_Other_Colors',
       'color_Patterned', 'color_Red/Orange', 'color_White',
       'color_Yellow/Gold/Cream'],
      dtype='object', length=122)


In [59]:
#Verificar o resultado no dataset Normalizado
print("Colunas após One-Hot Encoding (normalized):")
print(data_normalized.columns)

Colunas após One-Hot Encoding (normalized):
Index(['age_upon_outcome', 'animal_id_outcome', 'date_of_birth',
       'age_upon_outcome_(days)', 'age_upon_outcome_(years)',
       'age_upon_outcome_age_group', 'outcome_datetime', 'outcome_month',
       'outcome_year', 'outcome_monthyear',
       ...
       'is_mix_breed', 'processed_color', 'color_Black',
       'color_Brown/Chocolate', 'color_Gray/Blue', 'color_Other_Colors',
       'color_Patterned', 'color_Red/Orange', 'color_White',
       'color_Yellow/Gold/Cream'],
      dtype='object', length=122)


In [64]:
# Exibir todas as colunas dataset Padronizado
print("Colunas do dataset padronizado:")
for col in data_scaled.columns:
    print(col)

print(f"\nTotal de colunas no dataset padronizado: {len(data_scaled.columns)}")




Colunas do dataset padronizado:
age_upon_outcome
animal_id_outcome
date_of_birth
age_upon_outcome_(days)
age_upon_outcome_(years)
age_upon_outcome_age_group
outcome_datetime
outcome_month
outcome_year
outcome_monthyear
outcome_hour
outcome_number
dob_year
dob_month
dob_monthyear
age_upon_intake
animal_id_intake
breed
color
found_location
count
age_upon_intake_(days)
age_upon_intake_(years)
age_upon_intake_age_group
intake_datetime
intake_month
intake_year
intake_monthyear
intake_hour
intake_number
time_in_shelter
time_in_shelter_days
dob_year_scaled
age_upon_intake_(days)_scaled
age_upon_intake_(years)_scaled
intake_month_scaled
intake_year_scaled
intake_number_scaled
age_upon_outcome_(days)_scaled
age_upon_outcome_(years)_scaled
outcome_month_scaled
outcome_year_scaled
outcome_number_scaled
time_in_shelter_days_scaled
outcome_subtype_Aggressiv
outcome_subtype_At Vet   
outcome_subtype_Barn     
outcome_subtype_Behavior 
outcome_subtype_Court/Inv
outcome_subtype_Enroute  
outcome_subty

In [65]:
# Exibir todas as colunas dataset Normalizado
print("\nColunas do dataset normalizado:")
for col in data_normalized.columns:
    print(col)

print(f"\nTotal de colunas no dataset normalizado: {len(data_normalized.columns)}")


Colunas do dataset normalizado:
age_upon_outcome
animal_id_outcome
date_of_birth
age_upon_outcome_(days)
age_upon_outcome_(years)
age_upon_outcome_age_group
outcome_datetime
outcome_month
outcome_year
outcome_monthyear
outcome_hour
outcome_number
dob_year
dob_month
dob_monthyear
age_upon_intake
animal_id_intake
breed
color
found_location
count
age_upon_intake_(days)
age_upon_intake_(years)
age_upon_intake_age_group
intake_datetime
intake_month
intake_year
intake_monthyear
intake_hour
intake_number
time_in_shelter
time_in_shelter_days
dob_year_normalized
age_upon_intake_(days)_normalized
age_upon_intake_(years)_normalized
intake_month_normalized
intake_year_normalized
intake_number_normalized
age_upon_outcome_(days)_normalized
age_upon_outcome_(years)_normalized
outcome_month_normalized
outcome_year_normalized
outcome_number_normalized
time_in_shelter_days_normalized
outcome_subtype_Aggressiv
outcome_subtype_At Vet   
outcome_subtype_Barn     
outcome_subtype_Behavior 
outcome_subtype_

In [66]:
# Caminho para salvar os arquivos
scaled_file_path = '../data/scaled_ACC_INTAKES_OUTCOMES.parquet'
normalized_file_path = '../data/normalized_ACC_INTAKES_OUTCOMES.parquet'

# Salvar o dataset padronizado
data_scaled.to_parquet(scaled_file_path, engine='pyarrow')
print(f"Dataset padronizado salvo em: {scaled_file_path}")

# Salvar o dataset normalizado
data_normalized.to_parquet(normalized_file_path, engine='pyarrow')
print(f"Dataset normalizado salvo em: {normalized_file_path}")

Dataset padronizado salvo em: ../data/scaled_ACC_INTAKES_OUTCOMES.parquet
Dataset normalizado salvo em: ../data/normalized_ACC_INTAKES_OUTCOMES.parquet
